### PR TITLE
Use native HTTP monitors for HTTP(S) healthchecks

### PR DESCRIPTION
**Pull request by viranch**
Ran into a show stopper problem with TCP monitors: NS1 errors out if the health-check path contains `&`, usually shows up when using query params. Root cause unknown, but HTTP monitors don't seem to have that problem.

Plus using HTTP monitors is the more appropriate way of implementing HTTP(s) healthchecks anyway.

This change will cause sync to trigger a (graceful) update for all dynamic records that use HTTP healthchecks.

### CODE REVIEW

**Comment by ross** [comment]
/cc https://github.com/octodns/octodns/issues/713 for related discussion. Namely that NS1 pushed me towards using TCP this way when we run into problems during the initial development trying to use HTTP and unfortunately I've lost access to the work email that had the history/details as to why. 

Vaguely remember it having something to do with TLS validation, maybe HTTP didn't allow things to ignore self signed certs or something like that. It was 5-6 years ago and I really don't remember though :-(

**Comment by viranch** [comment]
Without knowing the original matter, NS1's current recommendation is to use HTTP monitors.

TLS validation is toggle-able in both types of monitors today, perhaps whatever it was is no more a standing issue.

I think we should be forward looking and use the appropriate monitor types for long term stability rather than patching TCP monitors to do HTTP, especially if the historical reason is no longer known and given there's a concrete issue at hand with TCP monitors.

**Comment by viranch** [comment]
> something to do with TLS validation

Perhaps it was to do with SNI? HTTP monitors will use the value of host header as SNI for SSL handshake. TCP monitors don't use any SNI.

**Comment by ross** [comment]
> I think we should be forward looking and use the appropriate monitor types for long term stability rather than patching TCP monitors to do HTTP, especially if the historical reason is no longer known and given there's a concrete issue at hand with TCP monitors.

Generally don't disagree, but it's a fairly dangerous area to mess with in terms of a deployment going out without any meaningful changes turning into an all health checks start failing and everything is going to the non-heath checked defaults. 

I do remember the change happening very early on, when I was probably only dealing with github so whatever it was was something that affected github. Unfortunately I don't have the visibility to test it there as thoroughly as I used to.

> Perhaps it was to do with SNI? HTTP monitors will use the value of host header as SNI for SSL handshake. TCP monitors don't use any SNI. But I believe that also shouldn't matter if TLS verify is disabled.

That's possible. If that was it things didn't use the host header for SNI at the time though.

**Comment by viranch** [comment]
Do you think a changelog entry plus a log.warn would be enough notice to users?

**Comment by viranch** [comment]
I realized that NS1 doesn't allow updating monitor types, so existing TCP monitors for HTTP health-checking will fail to get updated to HTTP type. I've added a commit for doing delete+create for this, but I think its now a breaking-enough change (answers go into a limbo until the new monitor status is populated) that should go behind a feature flag. That way users can opt out of this entirely if they want to stick to TCP monitors. How does that sound?

Btw I also realized this won't be a forward-compatible change, once users upgrade and run sync, they can't downgrade.

**Comment by ross** [comment]
Yeah opting in makes it a non-sensitive change. Probably wouldn't hurt to have a warning if this is the way you think things should go for new users. Down the road we could consider making the warn and error to force people to make the call between the old and new and explicitly set a value to get the existing TCP behavior.

Will try and sit down and look at the code in detail tomorrow.

File: octodns_ns1/__init__.py [hunk 1]
```diff
@@ -430,15 +431,21 @@ def __init__(
         self.log.debug(
             '__init__: id=%s, api_key=***, retry_count=%d, '
             'monitor_regions=%s, parallelism=%s, client_config=%s',
+            'shared_notifylist=%s, use_http_monitors=%s',
+            'default_healthcheck_http_version=%s',
```
Comment (ross) [inline octodns_ns1/__init__.py:435]: Good to get these added since they were missing.

**Review by ross — APPROVED** [review]

File: octodns_ns1/__init__.py [hunk 2]
```diff
@@ -1569,16 +1619,36 @@ def _extra_changes(self, desired, changes, **kwargs):
...
                     if not have:
-                        self.log.info(
-                            '_extra_changes: missing monitor %s', name
-                        )
+                        if self.use_http_monitors:
+                            self.log.warning(
+                                '_extra_changes: missing monitor %s will be created of type http, '
+                                'this will not be forward-compatible and octodns-ns1 cannot be downgraded',
+                                name,
+                            )
```
Comment (viranch) [inline octodns_ns1/__init__.py:1627]: here

File: octodns_ns1/__init__.py [hunk 3]
```diff
@@ -1569,16 +1619,36 @@ def _extra_changes(self, desired, changes, **kwargs):
...
+                            # delete+create, which has a few implications:
+                            self.log.warning(
+                                '_extra_changes: existing %s monitor %s will be deleted and replaced by a new %s monitor, '
+                                '%s will be temporarily treated as being healthy as a result '
+                                'and this is operation will be irreversible and not forward-compatible',
+                                have['job_type'],
+                                name,
+                                expected['job_type'],
+                                value,
+                            )
```
Comment (viranch) [inline octodns_ns1/__init__.py:1651]: and here

**Review by viranch — COMMENTED** [review]
@ross can you review the verbiage below?

File: octodns_ns1/__init__.py (hunk 2 above)
Comment (ross) [reply octodns_ns1/__init__.py:1627]: Seems fine as-is, optionally maybe throw an `' after applying'` on the end of it.

File: octodns_ns1/__init__.py (hunk 3 above)
Comment (ross) [reply octodns_ns1/__init__.py:1651]: Guess not super clear to me what's meant by the forward-compatible bit there. Does it mean that it won't work with older version of octodns-ns1 after applying? If so maybe just say that, similar to above.

File: octodns_ns1/__init__.py [hunk 4]
```diff
@@ -1569,16 +1619,37 @@ def _extra_changes(self, desired, changes, **kwargs):
...
+                            # delete+create, which has a few implications:
+                            self.log.warning(
+                                '_extra_changes: existing %s monitor %s will be deleted and replaced by a new %s monitor, '
+                                '%s will be temporarily treated as being healthy as a result, '
+                                'this is operation will be irreversible and not forward-compatible, ie '
+                                'octodns-ns1 cannot be downgraded below v0.0.5 after applying this change',
+                                have['job_type'],
+                                name,
+                                expected['job_type'],
+                                value,
```
Comment (viranch) [inline octodns_ns1/__init__.py:1651]: Starting to become a bit too wordy/lengthy, but I guess it sufficiently warns the user of all implications. LMK what you think @ross?

Comment (ross) [reply octodns_ns1/__init__.py:1651]: Works for me.
